crates/notion-core/src/shell/mod.rs

```diff
@@ -73,24 +60,16 @@ impl Shell for CurrentShell {
     }
 }
 
-/// Thrown when the shell name specified in the Notion environment is not supported.
-#[derive(Debug, Fail, NotionFail)]
-#[fail(display = "Unrecognized command shell name: {}", name)]
-#[notion_fail(code = "EnvironmentError")]
-struct UnrecognizedShellError {
-    name: String,
-}
-
 impl FromStr for CurrentShell {
     type Err = NotionError;
 
     fn from_str(src: &str) -> Result<Self, NotionError> {
-        let postscript_path = env::postscript_path().ok_or(UnspecifiedPostscriptError)?;
+        let postscript_path = env::postscript_path().ok_or(ErrorDetails::UnspecifiedPostscript)?;
 
         Ok(CurrentShell(match src {
             "bash" => Box::new(Bash { postscript_path }),
             _ => {
-                throw!(UnrecognizedShellError {
+                throw!(ErrorDetails::UnrecognizedShell {
```

Comment:
While we're doing the work to include the changes made for adding fish support here, we should tweak this implementation: wrapping a `throw!` *inside* an `Ok(...)` is super confusing, at least to me. It took me a few to figure out exactly what I was seeing. The actual semantics here, and therefore, what I'd recommend, are:

```rust
fn from_str(src: &str) -> Result<Self, NotionError> {
    let postscript_path = env::postscript_path().ok_or(ErrorDetails::UnspecifiedPostscript)?;

    match src {
        "bash" => Ok(CurrentShell(Box::new(Bash { postscript_path }))),
        "fish" => Ok(CurrentShell(Box::new(Fish { postscript_path }))),
        _ => Err(ErrorDetails::UnrecognizedShell {
            name: src.to_string(),
        }
        .into()),
    }
}
```

If there's a preference for keeping `throw!` instead of using `.into()`, you can do this:

```rust
fn from_str(src: &str) -> Result<Self, NotionError> {
    let postscript_path = env::postscript_path().ok_or(ErrorDetails::UnspecifiedPostscript)?;

    match src {
        "bash" => Ok(CurrentShell(Box::new(Bash { postscript_path }))),
        "fish" => Ok(CurrentShell(Box::new(Fish { postscript_path }))),
        _ => throw!(ErrorDetails::UnrecognizedShell {
            name: src.to_string(),
        }),
    }
}
```

Personally, I find that less clear, simply because there's such a nice symmetry with `Ok` and `Err` in the basic version.

Reply:
Agreed, that's a lot more explicit about the flow.